### PR TITLE
Tests for: magento/magento2#25585, magento/magento2#25248, magento/magento2#25058. 

### PR DESCRIPTION
**Pull request by p-bystritsky**
### Description (*)

Tests for: magento/magento2#25585, magento/magento2#25248, magento/magento2#25058. 

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
All the repetitive tasks should be extracted from Tests into Action Groups.

The main recommendation is that the Test should contain mostly Action Groups (with small exceptions)

File: app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            <deleteData createDataKey="createProductAttribute" stepKey="deleteProductAttribute"/>
+            <actionGroup ref="logout" stepKey="logout"/>
+        </after>
+        <!--Open product page-->
+        <amOnPage url="{{StorefrontProductPage.url($$createProductDefault.name$$)}}" stepKey="goToProductDefaultPage"/>
+        <waitForPageLoad stepKey="waitForProductPage"/>
+        <!--Click on 'Add to Compare' link-->
+        <waitForElementVisible selector="{{StorefrontProductInfoMainSection.productAddToCompare}}"
+                               stepKey="seeAddToCompareLink"/>
+        <click selector="{{StorefrontProductInfoMainSection.productAddToCompare}}" stepKey="clickOnAddToCompare"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml]: This section could be replaced - for example - with `StorefrontAddProductToCompareActionGroup`

File: app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        <waitForPageLoad stepKey="waitForProdAddToCmpList"/>
+        <!--Assert success message-->
+        <grabTextFrom selector="{{AdminProductMessagesSection.successMessage}}" stepKey="grabTextFromSuccessMessage"/>
+        <assertEquals expected='You added product $$createProductDefault.name$$ to the comparison list.' expectedType="string"
+                      actual="($grabTextFromSuccessMessage)" stepKey="assertSuccessMessage"/>
+        <!--See product in the comparison list-->
+        <amOnPage url="{{StorefrontProductComparePage.url}}" stepKey="navigateToComparePage"/>
+        <waitForPageLoad stepKey="waitForStorefrontProductComparePageLoad"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml]: And this one with `StorefrontOpenAndCheckComparisionActionGroup`

File: app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        <waitForPageLoad stepKey="waitForProductPage"/>
+        <!--Click on 'Add to Compare' link-->
+        <waitForElementVisible selector="{{StorefrontProductInfoMainSection.productAddToCompare}}"
+                               stepKey="seeAddToCompareLink"/>
+        <click selector="{{StorefrontProductInfoMainSection.productAddToCompare}}" stepKey="clickOnAddToCompare"/>
+        <waitForPageLoad stepKey="waitForProdAddToCmpList"/>
+        <!--Assert success message-->
+        <grabTextFrom selector="{{AdminProductMessagesSection.successMessage}}" stepKey="grabTextFromSuccessMessage"/>
+        <assertEquals expected='You added product $$createProductDefault.name$$ to the comparison list.' expectedType="string"
+                      actual="($grabTextFromSuccessMessage)" stepKey="assertSuccessMessage"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml]: This section could be replaced - for example - with `StorefrontAddProductToCompareActionGroup`

File: app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        <waitForPageLoad stepKey="waitForProductCustomPage"/>
+        <!--Click on 'Add to Compare' link-->
+        <waitForElementVisible selector="{{StorefrontProductInfoMainSection.productAddToCompare}}"
+                               stepKey="seeAddToCompareCustomLink"/>
+        <click selector="{{StorefrontProductInfoMainSection.productAddToCompare}}" stepKey="clickOnAddToCompareCustom"/>
+        <waitForPageLoad stepKey="waitForProdCustomAddToCmpList"/>
+        <!--Assert success message-->
+        <grabTextFrom selector="{{AdminProductMessagesSection.successMessage}}" stepKey="grabTextFromSuccessMessageCustom"/>
+        <assertEquals expected='You added product $$createProductCustom.name$$ to the comparison list.' expectedType="string"
+                      actual="($grabTextFromSuccessMessageCustom)" stepKey="assertSuccessMessageCustom"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml]: `StorefrontAddProductToCompareActionGroup`

File: app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml (hunk 2 above)
Comment (engcom-Foxtrot) [reply app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml]: This Action Group is not applicable from the Product Page.

Comment (p-bystritsky) [reply app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml]: This Action Group is not applicable from the Product Page.

File: app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml (hunk 3 above)
Comment (p-bystritsky) [reply app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml]: Updated

File: app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml (hunk 4 above)
Comment (p-bystritsky) [reply app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml]: Updated

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-6377](https://jira.corp.magento.com/browse/ENGCOM-6377) has been created to process this Pull Request
:eight_spoked_asterisk: @VladimirZaets, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

File: app/code/Magento/Catalog/Test/Mftf/ActionGroup/StorefrontCompareActionGroup.xml [hunk 5]
```diff
@@ -32,6 +32,8 @@
             <argument name="productVar"/>
         </arguments>
 
+        <waitForElementVisible selector="{{StorefrontProductInfoMainSection.productAddToCompare}}"
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/ActionGroup/StorefrontCompareActionGroup.xml]: New line break is not necessary

File: app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        <after>
+            <deleteData createDataKey="createProductDefault" stepKey="deleteProductDefault"/>
+            <deleteData createDataKey="createProductCustom" stepKey="deleteProductCustom"/>
+            <deleteData createDataKey="createCategory" stepKey="deleteCategory"/>
+            <deleteData createDataKey="createAttributeSet" stepKey="deleteAttributeSet"/>
+            <deleteData createDataKey="createProductAttribute" stepKey="deleteProductAttribute"/>
+            <actionGroup ref="logout" stepKey="logout"/>
+        </after>
+        <!--Open product page-->
+        <amOnPage url="{{StorefrontProductPage.url($$createProductDefault.name$$)}}" stepKey="goToProductDefaultPage"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml:49]: I still stand at the point you should use Action group there with argument passed.

File: app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml (hunk 2 above)
Comment (lbajsarowicz) [reply app/code/Magento/Catalog/Test/Mftf/Test/ProductAttributeWithoutValueInCompareListTest.xml]: Consider looking for other, creating your own

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Test should contain mostly ActionGroups.

**Comment by p-bystritsky** [comment]
@lbajsarowicz updated.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
You still need to extract repetitive actions to ActionGroups.

**Comment by p-bystritsky** [comment]
@lbajsarowicz I don't know what else can be extracted to action groups.

**Review by lbajsarowicz — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-6377](https://jira.corp.magento.com/browse/ENGCOM-6377) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
**Notice:** QA not applicable

**Comment by VladimirZaets** [comment]
@p-bystritsky Thanks. The main code is merged. Tests were updated by myself.

**Pull request closed without merging**